### PR TITLE
Fixed Administrative tables are too wide for smaller screens.

### PR DESCRIPTION
**Pull request by adamtothemoon**
pr for https://github.com/backdrop/backdrop-issues/issues/207

(take 2)

### CODE REVIEW

File: core/misc/tabledrag.js [hunk 1]
```diff
@@ -35,6 +35,7 @@ Drupal.tableDrag = function (table, tableSettings) {
   var self = this;
 
   // Required object variables.
+  this.$table = $(table);
```
Comment (quicksketch) [inline core/misc/tabledrag.js]: This line doesn't seem necessary as nothing else in tabledrag.js is changed. Seems like we could drop it.

File: core/includes/common.inc [hunk 2]
```diff
@@ -7159,7 +7159,7 @@ function drupal_common_theme() {
       'variables' => array(),
     ),
     'table' => array(
-      'variables' => array('header' => NULL, 'rows' => NULL, 'attributes' => array(), 'caption' => NULL, 'colgroups' => array(), 'sticky' => TRUE, 'empty' => ''),
+      'variables' => array('header' => NULL, 'rows' => NULL, 'attributes' => array(), 'caption' => NULL, 'colgroups' => array(), 'sticky' => TRUE,'responsive' => TRUE, 'empty' => ''),
```
Comment (quicksketch) [inline core/includes/common.inc]: Could you add a space between the comma and the first quote in `,'responsive' => TRUE`?

File: core/modules/node/node.admin.inc [hunk 3]
```diff
@@ -460,15 +460,33 @@ function node_admin_nodes() {
...
+  'status' => array(
+    'data' => t('Status'),
+    'field' => 'n.status',
+  ),
+  'changed' => array(
+    'data' => t('Updated'),
+    'field' => 'n.changed',
+    'sort' => 'desc',
+    'class' => array(RESPONSIVE_PRIORITY_LOW)
+   ,)
```
Comment (quicksketch) [inline core/modules/node/node.admin.inc]: The indentation here is all a little weird. Looks like it should all be indented 2 spaces. The last line here has the trailing comma in an odd place. It should be:

```
  'changed' => array(
    'data' => t('Updated'),
    'field' => 'n.changed',
    'sort' => 'desc',
    'class' => array(RESPONSIVE_PRIORITY_LOW),
  ),
```

File: core/modules/node/node.admin.inc [hunk 4]
```diff
@@ -460,15 +460,33 @@ function node_admin_nodes() {
...
   );
   if ($multilanguage) {
-    $header['language_name'] = array('data' => t('Language'), 'field' => 'n.langcode');
+   $header['language_name'] = array('data' => t('Language'), 'field' => 'n.langcode', 'class' => array(RESPONSIVE_PRIORITY_LOW));
```
Comment (quicksketch) [inline core/modules/node/node.admin.inc]: The replacement line should be the same indentation as the previous line (4 spaces total).

**Comment by quicksketch** [comment]
This in general looks great. It'll be fantastic to get this functionality into Backdrop. There a few concerns here that may need to be addressed in followups:
- I'm not sure about using ems for media queries. It doesn't seem like this was discussed in the original issue, but it doesn't seem to make a lot of sense. Even the code used as examples isn't consistent. Right now we have this code in multiple places:

```
+@media screen and (max-width:28.125em) { /* 450px */
```

But 28.125 ems isn't always going to be equal to 450 pixels; it depends on theme font size where the break point will actually be. Since Stark, Bartik, and Seven use different font sizes, it makes the break sizes at different pixel widths. This sort of misunderstanding around how ems work is why I prefer to avoid them entirely.
- We'll need to convert "Drupal" to "Backdrop" everywhere in this new code, but that can wait to be part of #3.

These issues aren't something we need to address in this PR, I just wanted to make note of them for future reference.

**Comment by adamtothemoon** [comment]
pr take 2

File: core/themes/bartik/css/style.css [hunk 5]
```diff
@@ -1490,3 +1489,21 @@ div.admin-panel .description {
 .js .dropbutton-multiple.open .dropbutton-widget:hover {
   background-color: white;
 }
+
+/**
+ * Responsive tables.
+ */
+@media screen and (max-width:28.125em) { /* 450px */
```
Comment (jenlampton) [inline core/themes/bartik/css/style.css]: looks like this is still in ems

File: core/themes/bartik/css/style.css [hunk 6]
```diff
@@ -1490,3 +1489,21 @@ div.admin-panel .description {
...
+  th.helpful,
+  td.helpful,
+  th.priority-medium,
+  td.priority-medium {
+    display: none;
+  }
+}
+@media screen and (max-width:45em) { /* 720px */
```
Comment (jenlampton) [inline core/themes/bartik/css/style.css]: this one too, also still in ems.

File: core/themes/seven/style.css [hunk 7]
```diff
@@ -341,12 +340,24 @@ ul.secondary li.active a.active {
...
-  margin-left: 40px; /* LTR */
   background: #fff;
-  position: relative;
   color: #333;
+  margin-left: 0.8125em;
+  margin-right: 0.8125em;
+  padding: 20px 0 40px 0;
+  position: relative;
+}
+@media screen and (min-width:28.125em) { /* 450px */
```
Comment (jenlampton) [inline core/themes/seven/style.css]: this one too, also still in ems.

File: core/themes/seven/style.css [hunk 8]
```diff
@@ -341,12 +340,24 @@ ul.secondary li.active a.active {
...
+  #page {
+    margin-left: 1.25em;
+    margin-right: 1.25em;
+  }
+}
+@media screen and (min-width:45em) { /* 720px */
```
Comment (jenlampton) [inline core/themes/seven/style.css]: this one too, also still in ems.

File: core/themes/seven/style.css [hunk 9]
```diff
@@ -528,6 +539,23 @@ tr td:last-child {
 
 
 /**
+ * Responsive tables.
+ */
+@media screen and (max-width:28.125em) { /* 450px */
```
Comment (jenlampton) [inline core/themes/seven/style.css]: this one too, also still in ems.

File: core/themes/seven/style.css [hunk 10]
```diff
@@ -528,6 +539,23 @@ tr td:last-child {
...
+  th.priority-low,
+  td.priority-low,
+  th.priority-medium,
+  td.priority-medium {
+    display: none;
+  }
+}
+@media screen and (max-width:45em) { /* 720px */
```
Comment (jenlampton) [inline core/themes/seven/style.css]: this one too, also still in ems.

File: core/themes/stark/layout.css [hunk 11]
```diff
@@ -143,3 +142,22 @@ main:after {
   }
 }
 
+/**
+ * Responsive tables.
+ */
+@media screen and (max-width:28.125em) { /* 450px */
```
Comment (jenlampton) [inline core/themes/stark/layout.css]: this one too, also still in ems.

File: core/themes/stark/layout.css [hunk 12]
```diff
@@ -143,3 +142,22 @@ main:after {
...
+  th.priority-low,
+  td.priority-low,
+  th.priority-medium,
+  td.priority-medium {
+    display: none;
+  }
+}
+@media screen and (max-width:45em) { /* 720px */
```
Comment (jenlampton) [inline core/themes/stark/layout.css]: this one too, also still in ems.

**Comment by jenlampton** [comment]
It looks like all these widths are still in ems and not in pixels.  Can you try again to change them?

**Comment by adamtothemoon** [comment]
I think I got them all

**Comment by jenlampton** [comment]
Looking good!  Assigning to quicksketch for final review and merge.

File: core/modules/system/system.module [hunk 13]
```diff
@@ -1152,6 +1152,20 @@ function system_library_info() {
     ),
   );
 
+  // Drupal's responsive table API.
+  $libraries['drupal.tableresponsive'] = array(
+    'title' => 'Drupal responsive table API',
+    'version' => VERSION,
```
Comment (quicksketch) [inline core/modules/system/system.module:1158]: VERSION should now be BACKDROP_VERSION. VERSION is the Drupal constant 7.x and only exists if backwards compatibility is enabled.

File: core/modules/system/system.module [hunk 14]
```diff
@@ -1152,6 +1152,20 @@ function system_library_info() {
...
+    'js' => array(
+      'core/misc/tableresponsive.js' => array('group' => JS_LIBRARY),
+    ),
+    'dependencies' => array(
+      array('system', 'jquery'),
+      array('system', 'drupal'),
+      array('system', 'jquery.once'),
```
Comment (quicksketch) [inline core/modules/system/system.module:1165]: We don't currently define jquery and drupal explicitly, they're added on all pages. So the two lines for jquery and drupal may be removed.

File: core/themes/seven/style.css [hunk 15]
```diff
@@ -341,12 +340,24 @@ ul.secondary li.active a.active {
...
+  #page {
+    margin-left: 1.25em;
+    margin-right: 1.25em;
+  }
+}
+@media screen and (min-width:720px) {
+  #page {
+    margin-left: 2.5em;
+    margin-right: 2.5em;
+  }
```
Comment (quicksketch) [inline core/themes/seven/style.css:360]: Let's stick with px margins here, which converts to:

```
0.8125em => 13px
1.25em => 20px
2.5em => 40px
```

File: core/misc/tableresponsive.js [hunk 16]
```diff
@@ -0,0 +1,141 @@
+(function ($, Drupal, window) {
```
Comment (quicksketch) [inline core/misc/tableresponsive.js:1]: The Drupal standards on which variables to include in the closure changed back and forth a couple of times. Eventually they decided that Drupal and window don't need to be scoped in closures, so we can remove `, Drupal, window` from this closure.

File: core/misc/tableresponsive.js [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+      this.$link.text(this.showText).data('pegged', 0);
+      // Refresh the toggle link.
+      $(window).trigger('resize.tableresponsive');
+    }
+  }
+});
+// Make the TableResponsive object available in the Drupal namespace.
+Drupal.TableResponsive = TableResponsive;
+
+})(jQuery, Drupal, window);
```
Comment (quicksketch) [inline core/misc/tableresponsive.js:141]: Same as the top of the file, `, Drupal, window` can be removed.

**Comment by quicksketch** [comment]
Overall this looks great and I'm excited about adding this functionality to Backdrop! Beside the additional feedback on the lines above, I also found that this has some inefficiencies in it.

When I first read this PR, I thought that the tableresponsive.js file was only added if the responsive classes were found. But as it turns out, this file is added for _every_ table unless specifically disabled. That means that all tables in Backdrop will get this JS file added even though they don't use the special classes. I think we should rework theme_table() so that the addition of this library is smart and automatic only if needed.

**Comment by quicksketch** [comment]
I filed a replacement PR at https://github.com/backdrop/backdrop/pull/311 based on this one that addresses the issues outlined above. Let's continue there.

**Pull request closed without merging**